metagpt/environment/base_env.py

```diff
@@ -185,7 +185,7 @@ def publish_message(self, message: Message, peekable: bool = True) -> bool:
         found = False
         # According to the routing feature plan in Chapter 2.2.3.2 of RFC 113
         for role, addrs in self.member_addrs.items():
-            if is_send_to(message, addrs):
+            if is_send_to(message, addrs) and role.is_watch(message.cause_by):
```

Comment:
- Only SOPs(standard operating procedure, like software team) use `cause_by`; 
- If it is a DI (`Data Interpreter`) or a `Generic Action`(have not implement yet), it can handle any message sent to it.
So this modification is not appropriate.

Reply:
Ok fair enough, If it breaks something then it is irrelevant to change.

Somethings which motivated me to do this changes:
- Lots of irrelevant message in `role.rc.msg_buffer`.
- Most of the time `role._observe` filters according to `cause_by` As here: https://github.com/geekan/MetaGPT/blob/bc1cd11a4b284e509056cc0052ec52d20568a4ab/metagpt/roles/role.py#L432

Don't get me wrong but I want to be clear on my side:
- I get it that you guys are using all messages in `Data Interpreter`. But you guys aren't using `Environment` in [Data Interpreter Role](https://github.com/geekan/MetaGPT/blob/main/metagpt/roles/di/data_interpreter.py) and I could not find anywhere in example of DI: https://github.com/geekan/MetaGPT/tree/main/examples/di

Reply:
Closing this pull, as this breaks Werewolf game: https://github.com/geekan/MetaGPT/tree/main/examples/werewolf_game

Reply:
The concept of "environment" is only needed when multiple agents coexist. The environment is responsible for establishing communication channels for them. 

The modules of MetaGPT can be adapted to different granularities of use cases. If you only need to tune the LLM, you can use the classes in the `metagpt/provider` folder; if you need a single agent, create a role object; only when you need multiple agents to collaborate do you need to create an "Environment" object, allowing the multiple agents to communicate with each other.